### ISSUE TITLE
Feature proposal: HTML, JS and CSS file autodiscovery

### ISSUE THREAD

**Issue by JuroOravec** [EmilStenstrom/django-components#380]
## TL;DR
Instead of explicitly listing all HTML, JS and CSS dependencies, add the option to allow `django-components` take care of it at runtime by either:
  - a) Looking for HTML/JS/CSS files in the directly where the python component file is
  - b) Looking for HTML/JS/CSS files in a different directory

---

## Description

Normally, we have to explicitly set the template `template_name`, as well as `Media.js` or `Media.css` files, like so:

```py
@component.register("button")
class Button(Component):
    template_name = "button.html"
    class Media:
        js = ["..."]
        css = ["..."]
```

In my project, the file paths are a lot longer, so the file declaration gets repetitive:

```py
@component.register("button")
class Button(Component):
    template_name = "module_name/components/component_name/component_name.html"
    class Media:
        js = "module_name/components/component_name/component_name.js"
        css = "module_name/components/component_name/component_name.css"
```

We also keep files for each component for in their own directory, e.g.:

```txt
app
|-- templates
    |--module_a
    |--module_b
        |-- pages
        |-- components
            |-- component_c
            |-- component_d
                |-- component_d.py
                |-- component_d.html
                |-- component_d.js
                |-- component_d.css
```

Thus, it would nice developer experience if we could skip listing all the files, and instead just do:

```py
@component.register("button")
class Button(Component):
	file_autodiscovery = True
```

As added bonus, this would make it easier to move the whole component folders around.

### Changing searched directory

In our use case, it's sufficient to look only in the parent directory of the component file. But for completeness, I think the feature should allow to change the directory which is searched for files.

For example, instead of searching in `app/templates/components/button`, following component would search for HTML, JS and CSS files in `app/templates/static/button`:

```py
@component.register("button")
class Button(Component):
	file_autodiscovery = True
    file_autodiscovery_root = "../static/button"
```

NOTE: To give users full flexibility, ideally they could specify 1) multiple dirs to be searched, and 2) set different dirs for each file type - HTML/CSS/JS. However, that's out of scope of this proposal.

## Implementation

Implementation makes use of a Metaclass that extends `django_components.component.SimplifiedInterfaceMediaDefiningClass`, which tries to pre-populate the `Component.template_name` and `Component.Media.js`/`Component.Media.css` attributes at class definition.

<details>
<summary> Click to expand  </summary>

```py
import os
from glob import glob
from typing import NamedTuple

from django.conf import settings

# Cache all available template dirs at startup
_template_dirs = [
    str(dir_path)
    for template_backend in (settings.TEMPLATES or [])
    for dir_path in template_backend.get("DIRS", [])
]


class TemplateDirResult(NamedTuple):
    absolute_dir: str
    template_root: str


def _get_template_dir_path_from_component_module_path(component_path):
    cls_path = component_path.replace(".", "/")
    cls_dir_path = os.path.dirname(cls_path)

    # Django resolves the template's relative to the values in TEMPLATE.DIRS
    # so we have to adjust the component dir path to match one of the template dirs.
    cwd = os.getcwd()
    absolute_dir_path = os.path.join(cwd, cls_dir_path)

    template_root = None
    for curr_dir in _template_dirs:
        if absolute_dir_path.startswith(curr_dir):
            # Get the common part of the path
            template_root = absolute_dir_path[: len(curr_dir)]
            break

    return TemplateDirResult(
        absolute_dir=absolute_dir_path, template_root=template_root
    )


class ComponentMetaclass(
    django_components.component.SimplifiedInterfaceMediaDefiningClass
):
    """
    Extension of the DjangoComponent's Component metaclass.

    We use this to automatically search for and include HTML, JS and CSS
    files that are in the same directory as the component class.

    To enable this, set `autodiscover_files` attribute to `True`.

    Even with autodiscovery, you can still add extra JS or CSS files
    via Media.js or Media.css, or define custom HTML file via `template_name`.
    """

    def __new__(mcs, name, bases, attrs):
        should_autodiscover = attrs.get("autodiscover_files")
        if should_autodiscover:
            should_search_html = not attrs.get("template_name")

            # Search for JS and CSS files in the same directory as where the component
            # is defined.
            cls_module_path = attrs["__module__"]
            result = _get_template_dir_path_from_component_module_path(cls_module_path)

            html_files = (
                glob(result.absolute_dir + "/*.html") if should_search_html else []
            )
            js_files = glob(result.absolute_dir + "/*.js")
            css_files = glob(result.absolute_dir + "/*.css")

            def trim_root(root: str):
                trimmed = root.replace(result.template_root, "")
                trimmed = trimmed[1:] if trimmed.startswith("/") else trimmed
                return trimmed

            # Make all found files to be relative to the closest Django template dir
            html_files = [trim_root(f) for f in html_files]
            js_files = [trim_root(f) for f in js_files]
            css_files = [trim_root(f) for f in css_files]

            # Unlike JS and CSS, there can be only 1 HTML file per component
            if should_search_html:
                if not len(html_files):
                    raise RuntimeError(
                        f"No HTML file found for component {name} in file {cls_module_path}"
                    )
                elif len(html_files) > 1:
                    raise RuntimeError(
                        f"Multiple HTML files found for component {name} in file {cls_module_path}."
                        " Cannot decide which one should be component's template"
                    )
                else:
                    # This is where we insert the auto-discovered HTML file into
                    # django-component's lifecycle.
                    attrs["template_name"] = html_files[0]

            media_cls = attrs.get("Media") or object
            media_js_files = media_cls.js or [] if hasattr(media_cls, "js") else []
            media_css_files = media_cls.js or [] if hasattr(media_cls, "css") else []

            # This is where we insert the auto-discovered JS and CSS files into
            # django-component's lifecycle.
            # NOTE: We extend the original Media cls if present in case the class defined
            # some funky behavior.
            new_media_cls = type(
                "media",
                (media_cls,),
                {
                    "js": [*media_js_files, *js_files],
                    "css": [*media_css_files, *css_files],
                },
            )

            attrs["Media"] = new_media_cls

        return super().__new__(mcs, name, bases, attrs)


class Component(django_components.component.Component, metaclass=ComponentMetaclass):
    autodiscover_files = False
```

</details>

**Comment by EmilStenstrom** [comment]
Hi @JuroOravec, thanks for the thorough proposal, and sorry for the late reply here.

There has been discussions about some kind of autodiscovery before. Long story short, I'm not a fan, because it makes understanding a component harder for new people. I really like the idea that you can look at the component class, and without knowing anything about how components work, find the related template, css and js files. Explicit over implicit.

This does does conflict with making components slim and fast to write, and I agree that the example you show doesn't look very good:

```python
@component.register("button")
class Button(Component):
    template_name = "module_name/components/component_name/component_name.html"
    class Media:
        js = "module_name/components/component_name/component_name.js"
        css = "module_name/components/component_name/component_name.css"
```

What I'd like to see (and I'm uncertain if it already works, I'd like to fix this if not!), is to allow relative paths to the files. so that this works:

```python
@component.register("button")
class Button(Component):
    template_name = "component_name.html"
    class Media:
        js = "component_name.js"
        css = "component_name.css"
```

Additionally, when we created the new startcomponent command, we recommended the more generic template.html, script.js, and style.css default names, so you wouldn't have to repeat yourself too much on that front either. So this is the recommended defaults now:

```python
@component.register("button")
class Button(Component):
    template_name = "template.html"
    class Media:
        js = "script.js"
        css = "style.css"
```

Comparing this to the below autodiscovery code, I don't think it's worth it to save a few lines of code by adding a layer of indirection.

```python
@component.register("button")
class Button(Component):
    file_autodiscovery = True
```

Of course, anyone not agreeing with the above sentiment, is open to specify their own base component class that acts differently.

**Comment by JuroOravec** [comment]
@EmilStenstrom Thanks for the reply! Agree with you that of the options above, being explicit but with relative paths would be ideal middle ground for clarity and convenience.

When I implemented this in our project, django-components (probably v0.30) worked such that the path had to be relative to the root of the Django's template dir, has this changed?

Either way, the implementation I posted at the end of the previous comment already includes the logic that, given a component class, resolves to a directory where the component class lives. Feel free to use it.

Moving on, if components allowed to import files via relative path, how would we distinguish between when a file is being imported relative to django's template dir vs relative to component file?
1. Would resolution relative to component file become the default? (breaking change)
2. Or would there be some kind of component or options flag to set the preferred way?
3. Or would it depend on the import path? E.g. if it starts with `./`, then it's a relative import, otherwise it's relative to django's template dir.
	- So `./script.js` is a relative import, whereas `script.js` is not.

**Comment by EmilStenstrom** [comment]
@JuroOravec To be honest, I don't know if we use relative paths or not. The ideal implementation would be that "script.js" looks for a local file, and falls back to global lookup if that fails. This is a breaking change, but I think it would be fairly safe, because there will often be a directory named like the component in between that makes absolute and relative paths different.

If possible, would you mind testing relative paths and opening a new PR if it doesn't work?

**Comment by JuroOravec** [comment]
Sure, I'll try to give it a go next weekend

**Comment by JuroOravec** [comment]
@EmilStenstrom I've started working on it in https://github.com/EmilStenstrom/django-components/pull/395, please give it a look when you will have the time

**Comment by JuroOravec** [comment]
Closing, the relative import feature was added in v0.60
